crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -972,13 +972,34 @@ impl CheckpointBuilder {
         pending: PendingCheckpointV2,
     ) -> anyhow::Result<()> {
         let pending = pending.into_v2();
+        let mut effects_in_current_checkpoint = BTreeSet::new();
+        let sorted_tx_effects_included_in_checkpoint = self
+            .resolve_checkpoint_transactions(pending.roots, &mut effects_in_current_checkpoint)
+            .await?;
+        let new_checkpoint = self
+            .create_checkpoints(sorted_tx_effects_included_in_checkpoint, pending.details)
+            .await?;
+        self.write_checkpoints(height, new_checkpoint).await?;
```

Comment:
The code is structure this way primarily due to easy integration with @aschran 's #17955 .

If this lends before that one, we just need to put 

```
let sorted_tx_effects_included_in_checkpoint = self
    .resolve_checkpoint_transactions(pending.roots, &mut effects_in_current_checkpoint)
    .await?;
```

into a loop over all the pending checkpoints, and concatenate all the `sorted_tx_effects_included_in_checkpoint`